Use NodeName and Namespace in CSS JIT

Constellation opened the pull request:
#### 6aefea97977e9843bddcd473b491014f7622795e
<pre>
Use NodeName and Namespace in CSS JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=267671">https://bugs.webkit.org/show_bug.cgi?id=267671</a>
<a href="https://rdar.apple.com/121167170">rdar://121167170</a>

Reviewed by Ryosuke Niwa and Justin Michaud.

1. Use NodeName in CSS JIT code generation. This is more efficient and we can reduce code size.
2. Drop Darwin ARMv7 CSS JIT since it is no longer used.
3. We adjust registers so that this patch removes weird configuration of Assemblers in CSS JIT (what registers are reserved etc.).

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::branch16):
(JSC::MacroAssemblerARM64::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::branch16):
(JSC::MacroAssemblerARMv7::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::branch16):
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::branch16):
(JSC::MacroAssemblerX86Common::branch8):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::branch16):
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementAttributeMatching):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementHasTagName):
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::QualifiedNameImpl::namespaceMemoryOffset):
(WebCore::QualifiedName::QualifiedNameImpl::nodeNameMemoryOffset):
(WebCore::QualifiedName::QualifiedNameImpl::namespaceURIMemoryOffset):

Canonical link: <a href="https://commits.webkit.org/276663@main">https://commits.webkit.org/276663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d4c8ebc29f3648e3f396c1c27b8092b075d4503

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45275 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24393 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47938 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41283 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21790 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45853 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21449 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39051 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18865 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40136 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3322 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/38495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41546 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49646 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44744 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20258 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16792 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44171 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21564 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42975 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51903 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6301 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21251 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10587 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/cssjit/RegisterAllocator.h:
```diff
@@ -84,7 +66,6 @@ static const JSC::MacroAssembler::RegisterID callerSavedRegisters[] = {
     JSC::X86Registers::r8,
     JSC::X86Registers::r9,
     JSC::X86Registers::r10,
-    JSC::X86Registers::r11
```
This is MacroAssembler scratch register. So removing it from this list.

Constellation commented on Source/WebCore/cssjit/RegisterAllocator.h:
```diff
@@ -100,7 +81,6 @@ static const JSC::MacroAssembler::RegisterID callerSavedRegisters[] = {
     JSC::X86Registers::r8,
     JSC::X86Registers::r9,
     JSC::X86Registers::r10,
-    JSC::X86Registers::r11
```
This is MacroAssembler scratch register. So removing it from this list.

Constellation commented on Source/WebCore/cssjit/RegisterAllocator.h:
```diff
@@ -212,11 +192,10 @@ class RegisterAllocator {
 #elif CPU(X86_64)
         return (registerID >= JSC::X86Registers::eax && registerID <= JSC::X86Registers::edx)
-            || (registerID >= JSC::X86Registers::esi && registerID <= JSC::X86Registers::r15);
+            || (registerID >= JSC::X86Registers::esi && registerID <= JSC::X86Registers::r10)
+            || (registerID >= JSC::X86Registers::r12 && registerID <= JSC::X86Registers::r15);
```
Removing r11.

rniwa commented on Source/WebCore/cssjit/SelectorCompiler.cpp:
```diff
@@ -3380,11 +3344,13 @@ void SelectorCodeGenerator::generateElementAttributeMatching(Assembler::JumpList
-                successCases.append(m_assembler.branchPtr(Assembler::Equal, Assembler::Address(qualifiedNameImpl, QualifiedName::QualifiedNameImpl::namespaceMemoryOffset()), namespaceToMatch));
+                if (attributeSelector.attribute().nodeNamespace() != Namespace::Unknown) {
+                    static_assert(sizeof(Namespace) == sizeof(uint8_t));
+                    successCases.append(m_assembler.branch8(Assembler::Equal, Assembler::Address(qualifiedNameImpl, QualifiedName::QualifiedNameImpl::namespaceMemoryOffset()), Assembler::TrustedImm32(static_cast<uint8_t>(attributeSelector.attribute().nodeNamespace()))));
```
why not enumToUnderlyingType?

Constellation replied:
Because `branch8` relies on that (not branch16, branch32, for example). So. we are explicitly putting static_assert and use `uint8_t` which is what the assembler want to see here.

rniwa commented on Source/WebCore/cssjit/SelectorCompiler.cpp:
```diff
@@ -4042,9 +4008,12 @@ inline void SelectorCodeGenerator::generateElementHasTagName(Assembler::JumpList
-            LocalRegister constantRegister(m_registerAllocator);
-            m_assembler.move(Assembler::TrustedImmPtr(selectorLocalName.impl()), constantRegister);
-            failureCases.append(m_assembler.branchPtr(Assembler::NotEqual, Assembler::Address(qualifiedNameImpl, QualifiedName::QualifiedNameImpl::localNameMemoryOffset()), constantRegister));
+            if (nameToMatch.nodeName() == NodeName::Unknown)
```
We should be consistent in using `== NodeName::Unknown` or `!= NodeName::Unknown`

Constellation replied:
It seems that this is the only place using this condition in if statement for now.

justinmichaud reviewed (approved):
r=me

Constellation committed:
Use NodeName and Namespace in CSS JIT
https://bugs.webkit.org/show_bug.cgi?id=267671
rdar://121167170

Reviewed by Ryosuke Niwa and Justin Michaud.

1. Use NodeName in CSS JIT code generation. This is more efficient and we can reduce code size.
2. Drop Darwin ARMv7 CSS JIT since it is no longer used.
3. We adjust registers so that this patch removes weird configuration of Assemblers in CSS JIT (what registers are reserved etc.).

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::branch16):
(JSC::MacroAssemblerARM64::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::branch16):
(JSC::MacroAssemblerARMv7::patchableBranch16):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::branch16):
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::branch16):
(JSC::MacroAssemblerX86Common::branch8):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::branch16):
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementAttributeMatching):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementHasTagName):
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::QualifiedNameImpl::namespaceMemoryOffset):
(WebCore::QualifiedName::QualifiedNameImpl::nodeNameMemoryOffset):
(WebCore::QualifiedName::QualifiedNameImpl::namespaceURIMemoryOffset):

Canonical link: https://commits.webkit.org/276663@main
